tests/state_test_util.go

```diff
@@ -156,11 +158,64 @@ func (t *StateTest) Subtests() []StateSubtest {
 	return sub
 }
 
-// Run executes a specific subtest.
-func (t *StateTest) Run(subtest StateSubtest, vmconfig vm.Config, isTestExecutionSpecState bool) (*state.StateDB, error) {
+// checkError checks if the error returned by the state transition matches any expected error.
+// A failing expectation returns a wrapped version of the original error, if any,
+// or a new error detailing the failing expectation.
+// This function does not return or modify the original error, it only evaluates and returns expectations for the error.
+func (t *StateTest) checkError(subtest StateSubtest, err error) error {
+	expectedError := t.json.Post[subtest.Fork][subtest.Index].ExpectException
+	if err == nil && expectedError == "" {
+		return nil
+	}
+	if err == nil && expectedError != "" {
+		return fmt.Errorf("expected error %q, got no error", expectedError)
+	}
+	if err != nil && expectedError == "" {
+		return fmt.Errorf("unexpected error: %w", err)
+	}
+	if err != nil && expectedError != "" {
+		// Ignore expected errors (TODO MariusVanDerWijden check error string)
+		return nil
+	}
+	return nil
+}
+
+func (t *StateTest) Run(subtest StateSubtest, vmconfig vm.Config, isTestExecutionSpecState bool) (result error) {
+	st, root, err := t.RunNoVerify(subtest, vmconfig, isTestExecutionSpecState)
+
+	// kaia-core-tests doesn't support ExpectException
+	if isTestExecutionSpecState {
+		checkedErr := t.checkError(subtest, err)
+		if checkedErr != nil {
+			return checkedErr
+		}
+	}
+
+	// The error has been checked; if it was unexpected, it's already returned.
+	if err != nil {
```

Comment:
Shouldn't it check `isTestExecutionSpecState`? Otherwise, `kaia-core-tests` tests will return nil on err.

```suggestion
	if isTestExecutionSpecState && err != nil {
```